src/Microsoft.Diagnostics.Runtime/src/DataReaders/MacOS/MacOSProcessDataReader.cs

```diff
@@ -0,0 +1,364 @@
+                if (kr != 0)
+                    if (kr != Native.KERN_INVALID_ADDRESS)
+                        throw new ClrDiagnosticsException();
+                    else
+                        break;
+
+                ulong endAddress = address + size;
+                result.Add(new MemoryRegion
+                {
+                    BeginAddress = address,
+                    EndAddress = endAddress,
+                    Permission = info.protection,
+                });
+
+                address = endAddress;
+            }
+
+            return result;
+        }
+
+        internal static class Native
+        {
+            internal const int EPERM = 1;
+
+            internal const int KERN_INVALID_ADDRESS = 1;
+
+            internal const int PROT_READ = 0x01;
+
+            internal const int PT_ATTACH = 10; // TODO: deprecated
+            internal const int PT_DETACH = 11;
+
+            internal const int TASK_DYLD_INFO = 17;
+            internal const int VM_REGION_BASIC_INFO_64 = 9;
+
+            internal static readonly unsafe int TASK_DYLD_INFO_COUNT = sizeof(task_dyld_info) / sizeof(uint);
+            internal static readonly unsafe int VM_REGION_BASIC_INFO_COUNT_64 = sizeof(vm_region_basic_info_64) / sizeof(int);
+
+            private const string LibSystem = "libSystem.dylib";
+
+            [DllImport(LibSystem, SetLastError = true)]
+            internal static extern int kill(int pid, int sig);
+
+            [DllImport(LibSystem)]
+            internal static extern int mach_task_self();
+
+            [DllImport(LibSystem, SetLastError = true)]
+            internal static extern int ptrace(int request, int pid, IntPtr addr = default, int data = default);
+
+            [DllImport(LibSystem)]
+            internal static extern int task_for_pid(int parent, int pid, out int task);
+
+            [DllImport(LibSystem)]
+            internal static extern int task_info(int target_task, uint flavor, out /*int*/task_dyld_info task_info, ref /*uint*/int task_info_count);
+
+            [DllImport(LibSystem)]
+            internal static extern unsafe int vm_read_overwrite(int target_task, /*UIntPtr*/ulong address, /*UIntPtr*/long size, /*UIntPtr*/void* data, out /*UIntPtr*/long data_size);
+
+            [DllImport(LibSystem)]
+            internal static extern int mach_vm_region(int target_task, ref /*UIntPtr*/ulong address, out /*UIntPtr*/ulong size, int flavor, out /*int*/vm_region_basic_info_64 info, ref /*uint*/int info_count, out int object_name);
+
+            [DllImport(LibSystem)]
+            internal static extern int waitpid(int pid, IntPtr status, int options);
+
+            internal readonly struct dyld_all_image_infos
+            {
+                internal readonly uint version;
+                internal readonly uint infoArrayCount;
+                internal readonly ulong infoArray;
+
+                // ...
```

Comment:
I assume we don't need the rest of this struct?

```suggestion
                // We don't need the rest of this struct so we do not define the rest of the fields.
```

Reply:
Yep the whole struct is too large.